nifi-commons/nifi-record/src/test/java/org/apache/nifi/serialization/record/ResultSetRecordSetTest.java

```diff
@@ -283,41 +544,235 @@ private ResultSet givenResultSetForArrayThrowsException(boolean featureSupported
+        assertEquals("The number of input fields does not match the number of fields in the result schema.", inputFields.size(), resultSchema.getFieldCount());
+        for (int i = 0; i < inputFields.size(); ++i) {
+            assertEquals(inputFields.get(i).getDataType(), resultSchema.getField(i).getDataType());
+        }
+    }
+
+    private void thenAllDataTypesAreString(final RecordSchema resultSchema) {
+        for (int i = 0; i < resultSchema.getFieldCount(); ++i) {
+            assertEquals(RecordFieldType.STRING.getDataType(), resultSchema.getField(i).getDataType());
+        }
     }
 
-    private void thenAllColumnDataTypesAreCorrect(final RecordSchema resultSchema) {
-        assertNotNull(resultSchema);
+    private void thenAllColumnDataTypesAreCorrect(TestColumn[] columns, RecordSchema expectedSchema, RecordSchema actualSchema) {
+        assertNotNull(actualSchema);
 
-        for (final Object[] column : COLUMNS) {
+        for (TestColumn column : columns) {
+            int fieldIndex = column.getIndex() - 1;
             // The DECIMAL column with scale larger than precision will not match so verify that instead
-            DataType actualDataType = resultSchema.getField((Integer) column[0] - 1).getDataType();
-            DataType expectedDataType = (DataType) column[3];
+            DataType actualDataType = actualSchema.getField(fieldIndex).getDataType();
+            DataType expectedDataType = expectedSchema.getField(fieldIndex).getDataType();
             if (expectedDataType.equals(RecordFieldType.DECIMAL.getDecimalDataType(3, 10))) {
                 DecimalDataType decimalDataType = (DecimalDataType) expectedDataType;
                 if (decimalDataType.getScale() > decimalDataType.getPrecision()) {
                     expectedDataType = RecordFieldType.DECIMAL.getDecimalDataType(decimalDataType.getScale(), decimalDataType.getScale());
                 }
             }
-            assertEquals("For column " + column[0] + " the converted type is not matching", expectedDataType, actualDataType);
+            assertEquals("For column " + column.getIndex() + " the converted type is not matching", expectedDataType, actualDataType);
+        }
+    }
+
+    private void thenActualArrayElementTypesMatchExpected(Map<String, DataType> expectedTypes, RecordSchema actualSchema) {
+        for (RecordField recordField : actualSchema.getFields()) {
+            if (recordField.getDataType() instanceof ArrayDataType) {
+                ArrayDataType arrayType = (ArrayDataType) recordField.getDataType();
+                if (!arrayType.getElementType().equals(expectedTypes.get(recordField.getFieldName()))) {
+                    throw new AssertionError("Array element type for " + recordField.getFieldName()
+                            + " is not of expected type " + expectedTypes.get(recordField.getFieldName()).toString());
+                }
```

Comment:
Is there a reason this cannot be changed to use `assertNotSame()` or perhaps `fail()` if one of the available assertions does not work?